Signal/src/network/PushManager.m

```diff
@@ -429,19 +431,61 @@ - (BOOL)applicationIsActive {
     return NO;
 }
 
-- (void)presentNotification:(UILocalNotification *)notification {
-    [[UIApplication sharedApplication] scheduleLocalNotification:notification];
-    [self.currentNotifications addObject:notification];
+- (void)presentNotification:(UILocalNotification *)notification checkForCancel:(BOOL)checkForCancel
+{
+    OWSAssert([NSThread isMainThread]);
+
+    NSString *threadId = notification.userInfo[Signal_Thread_UserInfo_Key];
+    if (checkForCancel && threadId != nil) {
+        [_pendingNotifications addObject:notification];
+
+        // The longer we wait, the more obsolete notifications we can suppress -
+        // but the more lag we introduce to notification delivery.
+        const CGFloat kDelaySeconds = 0.3f;
```

Comment:
Would it be simpler to modify the fireDate?

e.g. replace the entirety of this method implementation with:

```
- (void)presentNotification:(UILocalNotification *)notification checkForCancel:(BOOL)checkForCancel
{
    AssertIsOnMainThread();
    if (checkForCancel && threadId != nil) {
        _pendingNotification.fireDate = [NSDate dateWithTimeIntervalSinceNow:0.3f];
        _pendingNoficitation.timeZone = [NSTimeZone localTimeZone]; // IIRC this is necessary.
    }
    [[UIApplication sharedApplication] scheduleLocalNotification:notification];
    [self.currentNotifications addObject:notification];
}
```

Then most of the other changes could be omitted too.